src/Core/Scheduler/Providers/TileProvider.js

```diff
@@ -49,50 +41,84 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
     });
 };
 
+function defer() {
+    const deferedPromise = {};
+    deferedPromise.promise = new Promise((resolve, reject) => {
+        deferedPromise.resolve = resolve;
+        deferedPromise.reject = reject;
+    });
+    return deferedPromise;
+}
+
 TileProvider.prototype.executeCommand = function executeCommand(command) {
-    var extent = command.extent;
+    const extent = command.extent;
     if (command.requester &&
         !command.requester.material) {
         // request has been deleted
         return Promise.reject(new CancelledCommandException(command));
     }
 
-    var parent = command.requester;
-
-
-    // build tile
-    var params = {
-        layerId: command.layer.id,
-        extent,
-        level: (command.level === undefined) ? (parent.level + 1) : command.level,
-        segment: 16,
-        materialOptions: command.layer.materialOptions,
-        disableSkirt: command.layer.disableSkirt,
-    };
+    const parent = command.requester;
+    const level = (command.level === undefined) ? (parent.level + 1) : command.level;
 
-    const geometry = new TileGeometry(params, command.layer.builder);
+    if (!this.cacheGeometry[level]) {
+        this.cacheGeometry[level] = new Map();
+    }
 
-    var tile = new TileMesh(geometry, params);
+    const ce = command.layer.getCommonGeometryExtent(extent);
+    const south = ce.south().toFixed(8);
+
+    // build geometry if doesn't exist
+    if (!this.cacheGeometry[level].get(south)) {
+        this.cacheGeometry[level].set(south, defer());
```

Comment:
Why is it necessary to use a Promise mechanism here (instead of storing the geometry directly in the cache)?

Reply:
As the commands are asynchronous many tiles may ask for geometry during construction

Reply:
Everything here is synchronous, so it is guaranteed that no other part of itowns (or any other js in this page really) will get executed between this line and line 83. So the deferred will actually never be used here and you can directly set the TileGeometry in the cacheGeometry, without all the promise boilerplate.

Reply:
However the commands are executed in asynchronous, many tiles can ask for geometry during construction.
I tested it didn't work

Reply:
Please be more specific. What does "during construction" mean?

>  I tested it didn't work

What did you test? What didn't work?

> However the commands are executed in asynchronous

The fact that commands execute asynchronously has nothing to do with the matter. You set a promise on line 76 and resolve it on line 83.  No code can execute between these 2 lines, whether synchronous or asynchronous. 

Have you removed the `then` logic on line 87 when you tested?

Reply:
Thanks @autra , I started this PR a long time ago, where commands was concurrent.
I remove the promise in provider